### DISCUSSION TITLE
2023/03/10/llm-inf/

### DISCUSSION THREAD

**Discussion by giscus** [xv44586/giscus#6] [Announcements]
# 2023/03/10/llm-inf/

NLP | Machine Learning | Developer

https://xv44586.github.io/2023/03/10/llm-inf/

**Comment by thsno02** [comment]
`FP16 Tensor Core peak dense throughputs: ` 这里对新手来说有点容易误解，补充说明一下：

+ 流式多处理器（Stream Multi-processor，SM）是构建整个GPU的核心模块（执行整个Kernel Grid），一个流式多处理器上一般同时运行多个线程块。一般 SM 包含数量固定的 Tensor Cores 和 CUDA Cores 。通常我们说一个 GPU 有 n 个 CUDA Cores ，m 个 Tensor Cores，这些 Cores 都是均匀分布在 SM 里面的，比如有 a 个 SM，那么每一个 SM 有 n/a 个 CUDA Cores 和 m/a 个 Tensor Cores 。
+ 一般来说，peak theoretical throughput of the entire GPU (which includes both CUDA cores and Tensor cores) = `Peak GPU TFLOPS = (Total number of SMs) * (Clock Speed) * (Operations per cycle per SM)`：

	+ Total number of SMs 指的是一共有多少个 SM 单元，A100 中是 108 个。
	+ Clock Speed 指的是一秒内 Tensor Core 能执行一个 cycle 的频率 。cycle 可以理解为一次计算。这是由 GPU 性能所决定的，A100 中是 1.41 GHz。
	+ Operations per cycle per SM 指的是一个 SM 在进行一次 cycle 计算的时候执行了多少次操作。根据图 *Multiply-add operations per clock per SM* 可以知道在不同精度下，可执行的 cycle 数量是不一致的。在 FP16 下，A100 每一个 SM 可以执行 1024 次 cycle。在深度学习中，可以将 multiply-add 作为一个 cycle，那么就有两次 operation，multiplication 和 addition，所以是 2。

综上，即可得到作者给出的计算公式。

参考1：https://docs.nvidia.com/deeplearning/performance/dl-performance-gpu-background/index.html
参考2：https://www.nvidia.com/content/dam/en-zz/Solutions/Data-Center/a100/pdf/nvidia-a100-datasheet-nvidia-us-2188504-web.pdf
参考3：https://chat.openai.com/share/440260cc-c204-4588-bc97-d4b8d1d11326
